File: frontend/src/features/activity/activityStatus.test.ts
import { expect, it } from 'vitest'
import { compactStatusLabel, statusMarkKind, statusShowsRetry, type StatusFlags } from './activityStatus'

const flags = (value: Partial<StatusFlags>): StatusFlags => ({
  offline: false, pending: false, busy: false, hasSnapshot: true, error: null, ...value,
})

it('replaces confirmation with Saving… and keeps Offline while busy', () => {
  expect(compactStatusLabel(flags({ busy: true }))).toBe('Saving…')
  expect(compactStatusLabel(flags({ offline: true, busy: true }))).toBe('Offline · Saving…')
})

it('hides healthy Synced and keeps attention as a chip', () => {
  expect(statusMarkKind(flags({}))).toBeNull()
  expect(statusMarkKind(flags({ pending: true }))).toBe('chip')
  expect(statusMarkKind(flags({ busy: true }))).toBe('chip')
})

it('treats a storage or connection failure as Unsynced chrome with Retry', () => {
  const failed = flags({ pending: true, error: 'Activity storage failed: Storage full' })
  expect(compactStatusLabel(failed)).toBe('Unsynced')
  expect(statusShowsRetry(failed)).toBe(true)
  expect(compactStatusLabel(flags({ error: 'Connection lost' }))).toBe('Unsynced')
  expect(statusShowsRetry(flags({ error: 'Connection lost' }))).toBe(true)
  expect(statusShowsRetry(flags({ offline: true }))).toBe(false)
})

File: android/app/src/test/java/com/timebox/android/ui/day/ActivityStatusTest.kt
package com.timebox.android.ui.day

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class ActivityStatusTest {
    private fun flags(
        offline: Boolean = false,
        pending: Boolean = false,
        busy: Boolean = false,
        hasSnapshot: Boolean = true,
        error: String? = null,
    ) = StatusFlags(offline, pending, busy, hasSnapshot, error)

    @Test
    fun compactSavingReplacesConfirmationAndKeepsOffline() {
        assertEquals("Saving…", compactStatusLabel(flags(busy = true)))
        assertEquals("Offline · Saving…", compactStatusLabel(flags(offline = true, busy = true)))
    }

    @Test
    fun healthySyncedDrawsNothingAndAttentionIsAChip() {
        assertNull(statusMarkKind(flags()))
        assertEquals("chip", statusMarkKind(flags(pending = true)))
        assertEquals("chip", statusMarkKind(flags(busy = true)))
    }

    @Test
    fun storageOrConnectionFailureLooksLikeUnsyncedWithRetry() {
        val failed = flags(pending = true, error = "Activity storage failed: Storage full")
        assertEquals("Unsynced", compactStatusLabel(failed))
        assertTrue(statusShowsRetry(failed))
        assertEquals("Unsynced", compactStatusLabel(flags(error = "Connection lost")))
        assertTrue(statusShowsRetry(flags(error = "Connection lost")))
        assertFalse(statusShowsRetry(flags(offline = true)))
    }
}
